sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+from sbi.types import Shape, TorchTransform
+
+
+class NeuralPosteriorEnsemble(NeuralPosterior):
+    r"""Wrapper class for bundling together different posterior instances into an
+    ensemble.
+
+    This class creates a posterior ensemble from a set of N different, already trained
+    posterior estimators $p_{i}(\theta|x_o)$, where $i \in \{i,...,N\}$.
+
+    It can wrap all posterior classes available in sbi and even a mixture of different
+    posteriors, i.e. obtained via SNLE and SNPE at the same time, since it only
+    provides a pass-through to the class-methods of each posterior in the ensemble. The
+    only constraint is, that the individual posteriors have the same prior.
+
+    So far `log_prob()`, `sample()` and `map()` functionality are supported.
+
+    Example:
+
+    ```
+    import torch
+    from joblib import Parallel, delayed
+    from sbi.examples.minimal import simple
+
+    ensemble_size = 10
+    posteriors = Parallel(n_jobs=-1)(delayed(simple)() for i in range(ensemble_size))
+
+    ensemble = NeuralPosteriorEnsemble(posteriors)
+    ensemble.set_default_x(torch.zeros((3,)))
+    ensemble.sample((1,))
+    ```
+
+    Attributes:
+        posteriors: List of the posterior estimators making up the ensemble.
+        num_components: Number of posterior estimators.
+        weights: Weight of each posterior distribution. If none are provided each
+            posterior is weighted with 1/N.
+        potential_fn: Potential function of the ensemble.
+        prior: Prior distribution that is the same for all posteriors. If it is not the
+            same, then `prior = None`.
+        device: Device which the component distributions sit on.
+        default_x: Used in `.sample(), .log_prob()` as default conditioning context.
+    """
+
+    def __init__(
+        self,
+        posteriors: List,
+        weights: Optional[Union[List[float], Tensor]] = None,
+        theta_transform: Optional[TorchTransform] = None,
+    ):
+        """
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+            weights: Assign weights to posteriors manually, otherwise they will be
+                weighted with 1/N.
+            theta_transform: If passed, this transformation will be applied during the
+            optimization.
+        """
+        self.posteriors = posteriors
+        self.num_components = len(posteriors)
+        self.weights = weights
+        self.potential_fn = None  # will be set when context is provided.
+
+        self.device = self.ensure_same_device(posteriors)
+
+        self.prior = self.common_prior()
+        if theta_transform is not None:
+            self.theta_transform = theta_transform
+        elif self.prior is not None:
+            self.theta_transform = mcmc_transform(self.prior, device=self.device)
+        else:
+            self.theta_transform = None
+
+    def common_prior(self) -> bool:
```

Comment:
Please rename the method to `ensure_same_prior()`

Reply:
See other comment:
> I thought it might be desirable to keep the ensemble usable for posteriors with different proposals/priors. Therefore, I return the ensemble prior as None, if the components do not have the same prior. Prior is technically not needed anyway since the sampling and evaluations are done on each component independently. This is also the reason why I did not call the function ensure_same_prior, as I do not enforce it.